### PR TITLE
1502 refactoring sheet name

### PR DESCRIPTION
**Pull request by georgeDaskalakis**

### CODE REVIEW

**Comment by georgeDaskalakis** [comment]
@msevestre converted it to draft, to remove some comments that I saw in the diff and fix the function that adds a sheet, since it is not necessary to have sheetName as an argument

**Comment by msevestre** [comment]
Ok let me know what I need to review. We should meet this today so that we can get a beta for next week

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+using System.Data;
+using System.Linq;
+using OSPSuite.Utility.Collections;
+
+namespace OSPSuite.Infrastructure.Import.Core
+{
+   /// <summary>
+   ///    a collection of excel dataSheet sheets
+   /// </summary>
+   public class DataSheetCollection : IEnumerable<DataSheet>
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:12]: Why do we have a IDataSheet and a DataSheet? Do we need the interface ? if yes, this should be a collection of IDataSheet probably. It not remove

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+   {
+      private readonly Cache<string, DataSheet> _dataSheets;
+
+      public DataSheetCollection()
+      {
+         _dataSheets = new Cache<string, DataSheet>(x => x.SheetName);
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:18]: initialize where you define
private readonly Cache<string, DataSheet> _dataSheets = new Cache<string, DataSheet>(x => x.SheetName);

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      }
+
+      public DataSheetCollection GetDataSheets(IReadOnlyList<string> sheetNames)
+      {
+         var sheets = new DataSheetCollection();
+
+         foreach (var sheetName in sheetNames)
+         {
+            if (_dataSheets.Contains(sheetName))
+               sheets.AddSheet(_dataSheets[sheetName]);
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:38]: ok this is very weird here. A Method called GET modifies the data. Very suspicious to me

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      {
+         _dataSheets = new Cache<string, DataSheet>(x => x.SheetName);
+      }
+
+      public void AddSheet(DataSheet sheet)
+      {
+         _dataSheets.Add(sheet);
+      }
+
+      public DataSheet GetDataSheet(string sheetName)
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:26]: Call it SheetByName or sthg like this.

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      {
+         return _dataSheets[sheetName];
+      }
+
+      public DataSheetCollection GetDataSheets(IReadOnlyList<string> sheetNames)
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:31]: GetOrCreate. This has side effects and is not expected. Why is this needed?

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+      public DataSheetCollection Filter(string filter)
+      {
+         var filteredDataSheets = new DataSheetCollection();
+         foreach (var key in _dataSheets.Keys)
+         {
+            var dt = _dataSheets[key].AsDataTable();
+            var dv = new DataView(dt);
+            dv.RowFilter = filter;
+            var list = new List<DataRow>();
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:63]: why do you need this variable here?

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+         {
+            var data = new DataSet();
+            data.AddData(format.Parse(sheetKeyValue.Value, columnInfos));
+            dataSets.Add(sheetKeyValue.Key, data);
+         }
+
+         return dataSets;
+      }
+
+      public IReadOnlyList<string> AddNotExistingSheets(DataSheetCollection existingDataSheets)
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:89]: A comment explaining what is returned would be good here. using the /// to explain the behavior of this methid

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      {
+         var sheetNames = new List<string>();
+         foreach (var sheet in _dataSheets)
+         {
+            if (existingDataSheets.Contains(sheet.SheetName))
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:94]: you could filter right away using a where (_dataSheets.Where(x=>!_existingDataSheets.COntains(xx))
Matter of preference tho

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+               continue;
+
+            _dataSheets.Add(sheet);
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:97]: I don't get it. You are adding the sheet into the same collection. Something is very off here I think

**Review by msevestre — CHANGES_REQUESTED** [review]

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs [hunk 10]
```diff
@@ -17,48 +16,52 @@ public CsvDataSourceFile(IImportLogger logger, ICsvSeparatorSelector csvSeparato
...
-               {
-                  { "", dataSheet }
-               };
-               return loadedData;
+               DataSheets.AddSheet(dataSheet);
             }
          }
          catch (Exception e)
          {
+            DataSheets = alreadyLoadedDataSheets;
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs:64]: this will always be empty here

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs [hunk 11]
```diff
@@ -1,14 +1,206 @@
...
+
+      public DataTable AsDataTable()
+      {
+         var resultTable = new DataTable();
+         var indexList = _headers.Select( h => h.Index);
+
+         // Add columns.
+         foreach (var header in _headers.Keys)
+         {
+            resultTable.AddColumn(header, typeof(string));
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:130]: use extension methods AddColumn<string>

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs [hunk 12]
```diff
@@ -1,14 +1,206 @@
...
+         }
+         else
+         {
+            maskFunction = (inputList) => inputList;
+         }
+
+         foreach (var itemList in _rawDataTable)
+         {
+            // ReSharper disable once CoVariantArrayConversion
+            resultTable.Rows.Add(maskFunction(itemList).ToArray());
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:147]: are you sure about this. I have seen cases where this creates crash at run time

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs [hunk 13]
```diff
@@ -1,14 +1,206 @@
...
+
+         // Add columns.
+         foreach (var header in _headers.Keys)
+         {
+            resultTable.AddColumn(header, typeof(string));
+         }
+
+         Func<IEnumerable<string>, IEnumerable<string>> maskFunction;
+
+         if (_rawDataTable.Count > 0 && (_headers.Count != _rawDataTable.First().Count))
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:135]: Explain this condition. I can't understand what it does😊

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs [hunk 14]
```diff
@@ -1,14 +1,206 @@
...
+            if (_rawDataTable[i].TrueForAll(IsEmpty))
+               _rawDataTable.RemoveAt(i);
+            else
+               break;
+         }
+      }
+
+      public string SheetName { get; set; }
+
+      private static bool IsEmpty(string s)
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:200]: why lol

File: src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs [hunk 15]
```diff
@@ -189,9 +182,9 @@ private void setDefaultMetaData()
 
       private void createSheetsForViewing()
       {
-         foreach (var sheet in _dataSourceFile.DataSheets.KeyValues)
+         foreach (var sheet in _dataSourceFile.DataSheets)
          {
-            _sheetsForViewing[sheet.Key] = sheet.Value.RawData.AsDataTable();
+            _sheetsForViewing[sheet.SheetName] = sheet.AsDataTable();
```
Comment (msevestre) [inline src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:187]: WE started to rename those methods from AsDataTable to ToDataTable. Let's do it here as well

File: src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs [hunk 16]
```diff
@@ -257,10 +250,10 @@ public void RefreshTabs()
 
       public void DisableImportedSheets()
       {
-         if (Sheets.Keys.Any(x => x == View.SelectedTab))
+         if (ImportedSheets.Any(x => x.SheetName == View.SelectedTab))
             View.DisableImportCurrentSheet();
 
-         if (Sheets.Keys.All(GetSheetNames().Contains) && GetSheetNames().Count == Sheets.Keys.Count())
+         if (ImportedSheets.All(sheet => GetSheetNames().Contains(sheet.SheetName)) && GetSheetNames().Count == ImportedSheets.Count())
```
Comment (msevestre) [inline src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:256]: GetSheetNames() is called twice. You probably want to cash it before and do the count first as opposed to the check

Comment (msevestre) [reply src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:256]: We also have a COntainsAll method. Not sure if this is what you need. TO investigate as it might not check for the count

**Review by msevestre — COMMENTED** [review]
so there is no more IUnformattedData?

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 2 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:18]: done

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 3 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:38]: but it doesnt. sheets that gets modified is the local variable that gets returned. we get a list of sheet names, and we add to sheets the ones we find in the collection. so the internal data does not get modified. Or am I missing something?

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 4 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:26]: done

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 5 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:31]: @msevestre again, maybe I am missing something here....let's talk about this

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 6 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:63]: removed

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 9 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:97]: yeap, this is wrong you are right. corrected it.

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 7 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:89]: done

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 8 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:94]: yeap, it's just 3 lines like that. thank you!

**Comment by georgeDaskalakis** [comment]
> so there is no more IUnformattedData?

actually no. DataSheet used to be just IUnformattedData + sheetName and the Cache we were using was Cache<string, IUnformattedData> , where string was teh sheetName. I think the implementation as well as the naming makes more sense now.

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs (hunk 11 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:130]: without specifying typeof(string)

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs (hunk 14 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs]: fixed

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 1 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:12]: interface removed

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs [hunk 17]
```diff
@@ -17,48 +16,52 @@ public CsvDataSourceFile(IImportLogger logger, ICsvSeparatorSelector csvSeparato
...
-      protected override Cache<string, DataSheet> LoadFromFile(string path)
+      protected override void LoadFromFile(string path)
       {
+         var separator = _csvSeparatorSelector.GetCsvSeparator(path);
+
+         //if separator selection dialog was cancelled, abort
+         if (!(separator is char separatorCharacter)) return;
+
+         //we keep a copy of the already loaded sheets, in case the reading fails
+         var alreadyLoadedDataSheets = DataSheets;
```
[the remark this answers is not in the document]
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs:27]: I will let DataSheet implement IClonable, and implement Clone() as 

return this.MemberwiseClone();

this should be OK , right?

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs (hunk 11 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs:113]: done

File: src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs [hunk 18]
```diff
@@ -78,17 +71,17 @@ public void OnCompletedMapping()
       public void ImportDataForConfirmation(string sheetName)
       {
          var sheets = new Cache<string, DataSheet>();
```
[the remark this answers is not in the document]
Comment (georgeDaskalakis) [reply src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:73]: done

File: src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs (hunk 15 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:187]: done

File: src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs (hunk 16 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Presentation/Presenters/Importer/ImporterDataPresenter.cs:256]: done

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs (hunk 12 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs]: used .ToArray<object>()explicitely, no warning now and it works

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs [hunk 19]
```diff
@@ -1,14 +1,206 @@
...
+         foreach (var header in _headers.Keys)
+         {
+            resultTable.AddColumn(header, typeof(string));
+         }
+
+         Func<IEnumerable<string>, IEnumerable<string>> maskFunction;
+
+         if (_rawDataTable.Count > 0 && (_headers.Count != _rawDataTable.First().Count))
+         {
+            maskFunction = (inputList) => inputList.Where((v, i) => indexList.Contains(i));
```
[the remark this answers is not in the document]
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs]: comment added

File: src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs (hunk 13 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/IDataSheet.cs]: explained in the comment

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs]: I changed this to a constructor and it works. clone was not working

**Comment by msevestre** [comment]
How can clone not be working? This is something we do all the time. Do you have the code for me to see?

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (msevestre) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs]: No this implementation is not ok. You need to implement your own clone method

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+   /// <summary>
+   ///    a collection of excel dataSheet sheets
+   /// </summary>
+   public class DataSheetCollection : IEnumerable<DataSheet>
+   {
+      private readonly Cache<string, DataSheet> _dataSheets = new Cache<string, DataSheet>(x => x.SheetName);
+
+      public DataSheetCollection (DataSheetCollection dataSheetCollection)
+      {
+         _dataSheets = dataSheetCollection._dataSheets.Copy();
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:19]: This might not work. Where is the copy method coming from?

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs [hunk 21]
```diff
@@ -17,48 +16,52 @@ public CsvDataSourceFile(IImportLogger logger, ICsvSeparatorSelector csvSeparato
...
-               {
-                  { "", dataSheet }
-               };
-               return loadedData;
+               DataSheets.AddSheet(dataSheet);
             }
          }
          catch (Exception e)
          {
+            DataSheets = alreadyLoadedDataSheets;
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs:64]: Shouldn't DataSheet ( the member ) be read-only?

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/ExcelDataSourceFile.cs [hunk 22]
```diff
@@ -15,44 +14,46 @@ public ExcelDataSourceFile(IImportLogger logger) : base(logger)
...
 
-               loadedData.Add(sheetName, dataSheet);
+               DataSheets.AddSheet(rawSheetData);
             }
-
-            return loadedData;
          }
          catch (Exception ex)
          {
+            DataSheets = alreadyLoadedDataSheets;
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/ExcelDataSourceFile.cs:56]: Shouldn't the Datasheet (member) be read-only

**Review by msevestre — CHANGES_REQUESTED** [review]
Almost there I believe. Quick fix that we can perform live tomorrow

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs (hunk 21 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs]: I removed the setter from the DataSheet property

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/ExcelDataSourceFile.cs (hunk 22 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/ExcelDataSourceFile.cs]: I removed the setter from the DataSheet property

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs (hunk 20 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs]: I changed this

**Comment by georgeDaskalakis** [comment]
> How can clone not be working? This is something we do all the time. Do you have the code for me to see?

I am not sure why, but it did not work I checked again. I implemented the IClonable interface and the in Clone() I was using the memberWiseCopy. but when the original DataSheetCollection got cleared, the cloned one qould also get cleared.

The workaround I have found in the last commit I have pushed is to creat a CopySheetsFrom() method. I have tested it and it works.

**Comment by georgeDaskalakis** [comment]
> Almost there I believe. Quick fix that we can perform live tomorrow

@msevestre yeap, if you think my commit from today si not working as expected, let us quickly talk about the fix :)

File: src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+            {
+               ds.AddRow(drv.Row.ItemArray.Select(c => c.ToString()));
+            }
+            filteredDataSheets.AddSheet(ds);
+         }
+
+         return filteredDataSheets;
+      }
+
+      public void CopySheetsFrom (DataSheetCollection dataSheetsToCopy)
```
Comment (msevestre) [inline src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:69]: DataSheetCollection Clone() {
   var clone = new DataSheetCollection();
   _dataSheets.Each(clone.Add)
  return clone
}

Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSheetCollection.cs:69]: done

File: src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs [hunk 24]
```diff
@@ -17,48 +16,53 @@ public CsvDataSourceFile(IImportLogger logger, ICsvSeparatorSelector csvSeparato
...
-      protected override Cache<string, DataSheet> LoadFromFile(string path)
+      protected override void LoadFromFile(string path)
       {
+         var separator = _csvSeparatorSelector.GetCsvSeparator(path);
+
+         //if separator selection dialog was cancelled, abort
+         if (!(separator is char separatorCharacter)) return;
+
+         //we keep a copy of the already loaded sheets, in case the reading fails
+         var alreadyLoadedDataSheets = new DataSheetCollection();
```
[the remark this answers is not in the document]
Comment (georgeDaskalakis) [reply src/OSPSuite.Infrastructure.Import/Core/DataSourceFileReaders/CsvDataSourceFile.cs:27]: done

**Comment by georgeDaskalakis** [comment]
@msevestre all code review changes finalized
